### PR TITLE
feat(dagster-dbt): support dbt-core 1.4.x

### PR DESCRIPTION
**Pull request by rexledesma**
### Summary & Motivation
Resolves https://github.com/dagster-io/dagster/issues/11867.

Support both `v7/dbt-core==v1.3.*` and `v8/dbt-core==v1.4.*`.

Some things that have changed:
- The manifest artifact (empty fields are now none, so we should be more defensive when parsing these fields):
  - `v7`: https://schemas.getdbt.com/dbt/manifest/v7/index.html
  - `v8`: https://schemas.getdbt.com/dbt/manifest/v8/index.html
- The structured logging schema has different top-level keys https://docs.getdbt.com/reference/events-logging

### How I Tested These Changes
bk

### CODE REVIEW

**Comment by rexledesma** [comment]
Current dependencies on/for this PR:
* master
  * **PR #11902** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/11902" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/11902?utm_source=stack-comment).

File: examples/project_fully_featured/setup.py [hunk 1]
```diff
@@ -23,10 +23,8 @@
         "dagster-pyspark",
         "dagster-slack",
         "dagster-postgres",
-        "dbt-core",
-        "dbt-duckdb",
+        "dbt-duckdb==1.3.1",  # dbt-duckdb is not yet compatible with dbt 1.4.*
```
Comment (sryza) [inline examples/project_fully_featured/setup.py:26]: Why are we pinning duckdb, not dbt?

File: python_modules/libraries/dagster-dbt/dagster_dbt/asset_defs.py [hunk 2]
```diff
@@ -83,10 +83,12 @@ def _select_unique_ids_from_manifest_json(
 ) -> AbstractSet[str]:
     """Method to apply a selection string to an existing manifest.json file."""
     try:
+        import dbt.flags as flags
         import dbt.graph.cli as graph_cli
         import dbt.graph.selector as graph_selector
         from dbt.contracts.graph.manifest import Manifest
         from dbt.graph import SelectionSpec
+        from dbt.graph.selector_spec import IndirectSelection
```
Comment (sryza) [inline python_modules/libraries/dagster-dbt/dagster_dbt/asset_defs.py:91]: Have we tested to make sure that this won't raise the CheckError when using dbt >=1.0.0, but < 1.4

**Review by sryza — COMMENTED** [review]
How can we verify this works / make sure it's not breaking existing functionality

File: python_modules/libraries/dagster-dbt/dagster_dbt/asset_defs.py (hunk 2 above)
Comment (OwenKephart) [reply python_modules/libraries/dagster-dbt/dagster_dbt/asset_defs.py:91]: the dbt_13X tox environment installs dbt 1.3, and runs all the same tests

File: examples/project_fully_featured/setup.py (hunk 1 above)
Comment (OwenKephart) [reply examples/project_fully_featured/setup.py:26]: once the dbt-duckdb adapter is updated we can remove the pin on dbt-duckdb, but if dbt-core comes first in this list then I think dbt-core will be pulled in as 1.4.x

so it's easier to just specify the pin on the package that's causing the problem

**Review by sryza — APPROVED** [review]

**Pull request merged**